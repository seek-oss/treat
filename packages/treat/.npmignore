node_modules
src
tests
tsconfig.json
*.tsbuildinfo