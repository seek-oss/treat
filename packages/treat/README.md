# 🍬 treat

## [Docs](https://seek-oss.github.io/treat/)

See the documentation at seek-oss.github.io/treat for more information about using treat.

## License

MIT.
