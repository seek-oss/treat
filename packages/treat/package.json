{
  "name": "treat",
  "version": "1.6.1",
  "description": "Themeable, statically extracted CSS-in-JS with near-zero runtime",
  "main": "dist/treat.cjs.js",
  "module": "dist/treat.esm.js",
  "browser": {
    "./dist/treat.cjs.js": "./dist/treat.browser.cjs.js",
    "./dist/treat.esm.js": "./dist/treat.browser.esm.js"
  },
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "adapter.ts",
      "webpack-plugin/index.js",
      "webpack-plugin/loader.js"
    ]
  },
  "sideEffects": false,
  "author": "SEEK",
  "license": "MIT",
  "repository": "https://github.com/seek-oss/treat/tree/master/packages/treat",
  "peerDependencies": {
    "webpack": "^4.30.0 || ^5.20.2"
  },
  "dependencies": {
    "@hapi/joi": "^17.1.1",
    "@types/object-hash": "^1.2.0",
    "autoprefixer": "^10.2.4",
    "bluebird": "^3.5.4",
    "chalk": "^4.1.0",
    "css-loader": "^5.0.2",
    "css-selector-parser": "^1.3.0",
    "cssnano": "^4.1.10",
    "csstype": "^3.0.6",
    "debug": "^4.3.1",
    "dedent": "^0.7.0",
    "eval": "^0.1.2",
    "javascript-stringify": "^2.0.0",
    "loader-utils": "^1.2.3",
    "lodash": "^4.17.11",
    "normalize-path": "^3.0.0",
    "object-hash": "^2.0.3",
    "postcss": "^8.2.5",
    "postcss-js": "^3.0.3",
    "style-loader": "^2.0.0",
    "virtual-resource-loader": "^1.0.0"
  },
  "devDependencies": {
    "@types/bent": "^7.3.2",
    "@types/hapi__joi": "^17.1.6",
    "@types/lodash": "^4.14.134",
    "@types/mini-css-extract-plugin": "^1.2.2",
    "bent": "^7.3.12",
    "classnames": "^2.2.6",
    "mini-css-extract-plugin": "^1.3.5",
    "webpack": "^5.20.2"
  }
}
