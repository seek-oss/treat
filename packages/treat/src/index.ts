export {
  createTheme,
  style,
  styleMap,
  styleTree,
  globalStyle,
  globalStyleSheet,
} from './builder';
export { resolveStyles } from './resolveStyles';
export { resolveClassName } from './resolveClassName';

export {
  Style,
  GlobalStyle,
  GlobalStyleSheet,
  CSSProperties,
  ThemeRef,
  ClassRef,
  TreatModule,
} from './types';
