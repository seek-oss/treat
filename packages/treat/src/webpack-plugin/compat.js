const webpack4 = {
  isModuleUsed: (_compilation, module) =>
    typeof module.used === 'boolean' ? module.used : true,
  getDependencyModule: (_compilation, dependency) => dependency.module,
  getModulePreOrderIndex: (_compilation, module) => module.index,
  setModulePreOrderIndex: (_compilation, module, index) =>
    (module.index = index),
  getModulePostOrderIndex: (_compilation, module) => module.index2,
  setModulePostOrderIndex: (_compilation, module, index) =>
    (module.index2 = index),
  getCGModulePreOrderIndex: (chunkGroup, module) =>
    chunkGroup.getModuleIndex(module),
  setCGModulePreOrderIndex: (chunkGroup, module, index) =>
    chunkGroup.setModuleIndex(module, index),
  getCGModulePostOrderIndex: (chunkGroup, module) =>
    chunkGroup.getModuleIndex2(module),
  setCGModulePostOrderIndex: (chunkGroup, module, index) =>
    chunkGroup.setModuleIndex2(module, index),
  isModuleInChunk: (_compilation, module, chunk) =>
    chunk.containsModule(module),
  getModifiedFiles: (watchCompiler) => {
    // watchCompiler.watchFileSystem.watcher is undefined in some node environments.
    // Allow fallback to watchCompiler.watchFileSystem.wfs
    // https://github.com/s-panferov/awesome-typescript-loader/commit/c7da9ac82d105cdaf9b124ccc4c130648e59168a
    const watcher =
      watchCompiler.watchFileSystem.watcher ||
      watchCompiler.watchFileSystem.wfs.watcher;

    return Object.keys(watcher.mtimes);
  },
  getModuleIssuer: (_compilation, module) => module.issuer,
};

const webpack5 = {
  isModuleUsed: (compilation, module) => {
    const exportsInfo = compilation.moduleGraph.getExportsInfo(module);

    return exportsInfo.isModuleUsed('main');
  },
  getDependencyModule: (compilation, dependency) =>
    compilation.moduleGraph.getModule(dependency),
  getModulePreOrderIndex: (compilation, module) =>
    compilation.moduleGraph.getPreOrderIndex(module),
  setModulePreOrderIndex: (compilation, module, index) =>
    compilation.moduleGraph.setPreOrderIndex(module, index),
  getModulePostOrderIndex: (compilation, module) =>
    compilation.moduleGraph.getPostOrderIndex(module),
  setModulePostOrderIndex: (compilation, module, index) =>
    compilation.moduleGraph.setPostOrderIndex(module, index),
  getCGModulePreOrderIndex: (chunkGroup, module) =>
    chunkGroup.getModulePreOrderIndex(module),
  setCGModulePreOrderIndex: (chunkGroup, module, index) =>
    chunkGroup.setModulePreOrderIndex(module, index),
  getCGModulePostOrderIndex: (chunkGroup, module) =>
    chunkGroup.getModulePostOrderIndex(module),
  setCGModulePostOrderIndex: (chunkGroup, module, index) =>
    chunkGroup.setModulePostOrderIndex(module, index),
  isModuleInChunk: (compilation, module, chunk) =>
    compilation.chunkGraph.isModuleInChunk(module, chunk),
  getModifiedFiles: (watchCompiler) =>
    watchCompiler.modifiedFiles ? Array.from(watchCompiler.modifiedFiles) : [],
  getModuleIssuer: (compilation, module) =>
    compilation.moduleGraph.getIssuer(module),
};

export default (isWebpack5) => {
  if (isWebpack5) {
    return webpack5;
  }
  return webpack4;
};
