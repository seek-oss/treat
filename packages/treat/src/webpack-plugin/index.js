import Promise from 'bluebird';
import partition from 'lodash/partition';
import chalk from 'chalk';

import store from './store';
import reIndexModules from './reIndexModules';
import TreatError from './TreatError';
import makeTreatCompiler from './treatCompiler';
import optionValidator from './optionValidator';
import { debug } from './utils';
import createCompat from './compat';

const isProductionLikeMode = (options) => {
  return options.mode === 'production' || !options.mode;
};

const trace = debug('treat:webpack-plugin');

const TWP = 'treat-webpack-plugin';

const makeOptionDefaulter = (prodLike) => (option, { dev, prod }) => {
  if (typeof option !== 'undefined') {
    return option;
  }

  return prodLike ? prod : dev;
};

export class TreatPlugin {
  constructor(options = {}) {
    optionValidator(options);

    const {
      test = /\.treat\.(js|ts)$/,
      outputCSS = true,
      outputLoaders = ['style-loader'],
      localIdentName,
      themeIdentName,
      minify,
      browsers,
      hmr,
    } = options;

    this.store = store();
    this.treatCompiler = makeTreatCompiler();

    this.test = test;
    this.minify = minify;
    this.localIdentName = localIdentName;
    this.themeIdentName = themeIdentName;
    this.loaderOptions = {
      outputCSS,
      outputLoaders,
      browsers,
      hmr,
    };
  }

  apply(compiler) {
    const isWebpack5 = Boolean(compiler.webpack && compiler.webpack.version);
    const compat = createCompat(isWebpack5);

    compiler.hooks.watchRun.tap(TWP, (watchCompiler) => {
      const modifiedFiles = compat.getModifiedFiles(watchCompiler);

      this.treatCompiler.expireCache(modifiedFiles);
    });

    compiler.hooks.thisCompilation.tap(TWP, (compilation) => {
      let allCssModules = [];
      let usedCssModules;

      const rebuildModule = (module) =>
        new Promise((resolve) => {
          compilation.rebuildModule(module, () => {
            resolve();
          });
        });

      compilation.hooks.finishModules.tapPromise(TWP, async () => {
        // Filter out css modules that no longer exist (watch mode)
        this.store
          .getAllCssOwners()
          .filter((identifier) => !Boolean(compilation.findModule(identifier)))
          .forEach((identifier) => {
            this.store.deleteCssRequests(identifier);
          });

        // Filter out theme treat modules that no longer exist (watch mode)
        this.store
          .getThemeIdentifiers()
          .filter((identifier) => !Boolean(compilation.findModule(identifier)))
          .forEach((identifier) => {
            this.store.deleteModuleThemes(identifier);
          });

        const finalThemeHash = this.store.getThemesHash();
        const builtTreatModules = this.store
          .popAllModules()
          .filter(
            ({ moduleIdentifier }) =>
              !this.store.getThemeIdentifiers().includes(moduleIdentifier),
          );

        // Rebuild all non theme treat modules that built before all themes were registered
        const staleModules = builtTreatModules.filter(
          ({ themeHash }) => themeHash != finalThemeHash,
        );

        await Promise.map(staleModules, async ({ moduleIdentifier }) => {
          trace('Rebuiling stale module: %i', moduleIdentifier);
          const treatModule = compilation.findModule(moduleIdentifier);

          if (!treatModule) {
            // In cases where the build fails, sometimes modules are not present
            // on the compilation object. Skip rebuilding in these cases.
            trace(
              "Bail on rebuild of stale module: %i Module doesn't exist on compilation",
              moduleIdentifier,
            );
            return;
          }

          await rebuildModule(treatModule);
        });

        // Modules should be upto date so clear rebuild queue
        this.store.popAllModules();

        if (builtTreatModules.length > 0) {
          // Rebuild all theme modules to ensure they contain upto date css
          await Promise.map(
            this.store.getThemeIdentifiers(),
            async (themeModuleIdentifier) => {
              trace('Rebuiling theme module: %i', themeModuleIdentifier);

              const treatModule = compilation.findModule(themeModuleIdentifier);

              if (!treatModule) {
                // In cases where the build fails, sometimes modules are not present
                // on the compilation object. Skip rebuilding in these cases.
                trace(
                  "Bail on rebuild of theme module: %i Module doesn't exist on compilation",
                  themeModuleIdentifier,
                );
                return;
              }

              await rebuildModule(treatModule);
            },
          );
        }

        const cssResources = this.store
          .getAllThemedCssRequests()
          .map(({ resource }) => resource);

        const getTreatCssResource = (module) => {
          if (
            module.matchResource &&
            cssResources.includes(module.matchResource)
          ) {
            return module.matchResource;
          }

          if (module.type === 'css/mini-extract') {
            const issuer = compat.getModuleIssuer(compilation, module);

            return getTreatCssResource(issuer);
          }

          return null;
        };

        for (const module of compilation._modules.values()) {
          const treatCssResource = getTreatCssResource(module);

          if (treatCssResource) {
            const { owner, type, theme } = this.store.getThemedCssModuleInfo(
              treatCssResource,
            );
            const themeIdentifier = theme
              ? this.store.getTheme(theme).identifier
              : null;

            allCssModules.push({
              type,
              module: compilation.findModule(module.identifier()),
              owner: compilation.findModule(owner),
              themeModule: theme
                ? compilation.findModule(themeIdentifier)
                : null,
              identifier: module.identifier(),
            });
          }
        }
      });

      compilation.hooks.optimizeDependencies.tap(
        // Running during advanced stage means export usage info has been added to the graph already
        { name: TWP, stage: 10 },
        () => {
          // Some modules may not be used and their css can be safely be removed from the chunks
          const cssModuleGroups = partition(
            allCssModules,
            ({ module, owner, themeModule }) => {
              const ownerIsUsed = compat.isModuleUsed(compilation, owner);
              const themeIsUsed =
                !themeModule || compat.isModuleUsed(compilation, themeModule);

              const cssModuleIsUsed = ownerIsUsed && themeIsUsed;

              if (!cssModuleIsUsed) {
                trace(
                  'CSS Module marked for removal due to unused %s %m',
                  chalk.yellow(ownerIsUsed ? 'theme' : 'owner'),
                  module,
                );
              }

              return cssModuleIsUsed;
            },
          );

          usedCssModules = cssModuleGroups[0];
          const cssModulesToRemove = cssModuleGroups[1];

          if (cssModulesToRemove.length > 0) {
            const themeModules = this.store
              .getThemeIdentifiers()
              .map((ident) => compilation.findModule(ident));

            const cssModulesIdentsToRemove = cssModulesToRemove.map(
              ({ identifier }) => identifier,
            );

            for (const themeModule of themeModules) {
              const depsToRemove = themeModule.dependencies.filter((dep) => {
                const depModule = compat.getDependencyModule(compilation, dep);

                const shouldRemove =
                  depModule &&
                  cssModulesIdentsToRemove.includes(depModule.identifier());

                if (shouldRemove) {
                  trace('Mark %m for clean-up in %m', depModule, themeModule);
                }

                return shouldRemove;
              });

              for (const dep of depsToRemove) {
                themeModule.removeDependency(dep);
              }

              compilation.removeReasonsOfDependencyBlock(themeModule, {
                dependencies: depsToRemove,
                blocks: [],
              });
            }
          }
        },
      );

      compilation.hooks.afterChunks.tap(TWP, (chunks) => {
        try {
          // afterChunks hook means module/chunk order properties have been set
          // We can now correct those values by referencing the ordering of the owner treat file
          for (const chunk of chunks) {
            const cssModulesInChunk = usedCssModules.filter(({ module }) =>
              compat.isModuleInChunk(compilation, module, chunk),
            );

            for (const chunkGroup of chunk.groupsIterable) {
              // Corrects ChunkGroup._moduleIndicies
              // Used for mini-extract-css-plugin ordering
              reIndexModules(
                cssModulesInChunk,
                {
                  getPreIndex: ({ module }) =>
                    compat.getCGModulePreOrderIndex(chunkGroup, module),
                  getPostIndex: ({ module }) =>
                    compat.getCGModulePostOrderIndex(chunkGroup, module),
                  getOwnerIndex: ({ owner }) =>
                    compat.getModulePreOrderIndex(compilation, owner),
                  getThemeIndex: ({ themeModule }) =>
                    compat.getModulePreOrderIndex(compilation, themeModule),
                  setPreIndex: ({ module }, i) =>
                    compat.setCGModulePreOrderIndex(chunkGroup, module, i),
                  setPostIndex: ({ module }, i) =>
                    compat.setCGModulePostOrderIndex(chunkGroup, module, i),
                },
                'ChunkGroup moduleIndicies',
              );
            }
          }

          // Corrects Module.index/index2
          reIndexModules(
            usedCssModules,
            {
              getPreIndex: ({ module }) =>
                compat.getModulePreOrderIndex(compilation, module),
              getPostIndex: ({ module }) =>
                compat.getModulePostOrderIndex(compilation, module),
              getOwnerIndex: ({ owner }) =>
                compat.getModulePreOrderIndex(compilation, owner),
              getThemeIndex: ({ themeModule }) =>
                compat.getModulePreOrderIndex(compilation, themeModule),
              setPreIndex: ({ module }, i) =>
                compat.setModulePreOrderIndex(compilation, module, i),
              setPostIndex: ({ module }, i) =>
                compat.setModulePostOrderIndex(compilation, module, i),
            },
            'Module.index/index2',
          );

          this.store
            .getThemeIdentifiers()
            .map((themeModuleIdentifier) =>
              compilation.findModule(themeModuleIdentifier),
            )
            .forEach((themeModule) => {
              const { dependencies } = themeModule;

              const relevantDependencies = usedCssModules
                .map((moduleInfo) => ({
                  ...moduleInfo,
                  dependency: dependencies.find((d) => {
                    const module = compat.getDependencyModule(compilation, d);

                    return (
                      module && module.identifier() === moduleInfo.identifier
                    );
                  }),
                }))
                .filter(({ dependency }) => dependency);

              // Corrects Dependency.sourceOrder
              // Used for regular import chunk sorting, e.g. style-loader
              reIndexModules(
                relevantDependencies,
                {
                  getPreIndex: ({ dependency }) => dependency.sourceOrder,
                  getOwnerIndex: ({ owner }) =>
                    compat.getModulePreOrderIndex(compilation, owner),
                  getThemeIndex: ({ themeModule }) =>
                    compat.getModulePreOrderIndex(compilation, themeModule),
                  setPreIndex: ({ dependency }, i) => {
                    dependency.sourceOrder = i;
                  },
                  // post index not required for dependencies
                  getPostIndex: () => {},
                  setPostIndex: () => {},
                },
                'Dependency.sourceOrder',
              );
            });
        } catch (e) {
          compilation.errors.push(new TreatError(e));
        }
      });
    });

    compiler.hooks.normalModuleFactory.tap(TWP, (nmf) => {
      nmf.hooks.afterResolve.tap(TWP, (result) => {
        if (this.store.getCSSResources().has(result.matchResource)) {
          result.settings = Object.assign({}, result.settings, {
            sideEffects: true,
          });
        }
      });
    });

    const optionDefaulter = makeOptionDefaulter(
      isProductionLikeMode(compiler.options),
    );

    compiler.options.module.rules.splice(
      0,
      0,
      {
        test: this.test,
        use: [
          {
            loader: require.resolve('treat/webpack-plugin/loader'),
            options: {
              ...this.loaderOptions,
              minify: optionDefaulter(this.minify, {
                dev: false,
                prod: true,
              }),
              localIdentName: optionDefaulter(this.localIdentName, {
                dev: '[name]-[local]_[hash:base64:5]',
                prod: '[hash:base64:5]',
              }),
              themeIdentName: optionDefaulter(this.themeIdentName, {
                dev: '_[name]-[local]_[hash:base64:4]',
                prod: '[hash:base64:4]',
              }),
              store: this.store,
              treatCompiler: this.treatCompiler,
            },
          },
        ],
      },
      {
        test: /\.treatcss$/,
        sideEffects: true,
        use: [
          ...this.loaderOptions.outputLoaders,
          {
            loader: 'css-loader',
            options: {
              modules: false,
              url: false,
            },
          },
        ],
      },
    );
  }
}
