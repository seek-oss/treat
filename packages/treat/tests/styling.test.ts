import MiniCssExtractPlugin from 'mini-css-extract-plugin';
import startFixture from 'treat-test-helpers/startServer';
import getStyles from 'treat-test-helpers/getStyles';
import { TreatPlugin } from 'treat/webpack-plugin';

const mainSelector = '#main';

const configVariations = [
  {
    label: 'style-loader',
    config: () => ({
      plugins: [new TreatPlugin()],
    }),
  },
  {
    label: 'mini-css-extract-loader',
    config: () => ({
      optimization: {
        splitChunks: {
          cacheGroups: {
            styles: {
              name: 'styles',
              type: 'css/mini-extract',
              chunks: 'all',
              enforce: true,
            },
          },
        },
      },
      plugins: [
        new TreatPlugin({
          outputLoaders: [MiniCssExtractPlugin.loader],
        }),
        new MiniCssExtractPlugin(),
      ],
    }),
  },
  {
    label: 'development mode (style-loader)',
    config: () => ({
      mode: 'development',
      plugins: [new TreatPlugin()],
    }),
  },
];

describe('Styling and specificity', () => {
  configVariations.forEach(({ label, config }) => {
    describe(`Simple - ${label} - `, () => {
      let server: any;

      beforeAll(async () => {
        server = await startFixture({
          entry: require.resolve('./fixtures/simple/index.ts'),
          ...config(),
        });
        await page.goto(server.url);
      });

      describe('on mobile', () => {
        beforeAll(async () => {
          await page.setViewport({ width: 400, height: 800 });
        });

        it('should be styled correctly', async () => {
          const styles = await getStyles(page, mainSelector);

          expect(styles).toMatchSnapshot();
        });
      });

      describe('on desktop', () => {
        beforeAll(async () => {
          await page.setViewport({ width: 1200, height: 900 });
        });

        it('should be styled correctly', async () => {
          const styles = await getStyles(page, mainSelector);

          expect(styles).toMatchSnapshot();
        });

        describe('and hovered', () => {
          beforeAll(async () => {
            await page.hover(mainSelector);
          });

          it('should be styled correctly', async () => {
            const styles = await getStyles(page, mainSelector);

            expect(styles).toMatchSnapshot();
          });
        });
      });

      afterAll(() => {
        server.close();
      });
    });

    describe(`Themed - ${label}`, () => {
      let server: any;

      beforeAll(async () => {
        server = await startFixture({
          entry: require.resolve('./fixtures/themed/index.ts'),
          ...config(),
        });
        await page.goto(server.url);
      });

      it('should be styled correctly with theme one', async () => {
        const styles = await getStyles(page, '#theme1');

        expect(styles).toMatchSnapshot();
      });

      it('should be styled correctly with theme two', async () => {
        const styles = await getStyles(page, '#theme2');

        expect(styles).toMatchSnapshot();
      });

      afterAll(() => {
        server.close();
      });
    });

    describe(`Dynamic imports - ${label}`, () => {
      let server: any;

      beforeAll(async () => {
        server = await startFixture({
          entry: require.resolve('./fixtures/dynamic-imports/index.ts'),
          ...config(),
        });
        await page.goto(server.url);
      });

      describe('on mobile', () => {
        beforeAll(async () => {
          await page.setViewport({ width: 400, height: 800 });
        });

        it('theme A should be styled correctly', async () => {
          const styles = await getStyles(page, '#themeA');

          expect(styles).toMatchSnapshot();
        });

        it('theme B should be styled correctly', async () => {
          const styles = await getStyles(page, '#themeB');

          expect(styles).toMatchSnapshot();
        });
      });

      afterAll(() => {
        server.close();
      });
    });

    describe(`Unused modules - ${label}`, () => {
      let server: any;

      beforeAll(async () => {
        server = await startFixture({
          entry: require.resolve('./fixtures/unused-modules/index.ts'),
          ...config(),
        });
        await page.goto(server.url);
      });

      it('should be styled correctly', async () => {
        const styles = await getStyles(page, mainSelector);

        expect(styles).toMatchSnapshot();
      });

      afterAll(() => {
        server.close();
      });
    });
  });
});
