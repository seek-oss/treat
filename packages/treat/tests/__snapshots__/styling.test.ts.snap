// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Styling and specificity Dynamic imports - development mode (style-loader) on mobile theme A should be styled correctly 1`] = `
Object {
  "background-color": "yellow",
  "display": "flex",
  "height": "100px",
  "width": "100px",
}
`;

exports[`Styling and specificity Dynamic imports - development mode (style-loader) on mobile theme B should be styled correctly 1`] = `
Object {
  "background-color": "red",
  "display": "flex",
  "height": "100px",
  "width": "100px",
}
`;

exports[`Styling and specificity Dynamic imports - mini-css-extract-loader on mobile theme A should be styled correctly 1`] = `
Object {
  "background-color": "#ff0",
  "display": "flex",
  "height": "100px",
  "width": "100px",
}
`;

exports[`Styling and specificity Dynamic imports - mini-css-extract-loader on mobile theme B should be styled correctly 1`] = `
Object {
  "background-color": "red",
  "display": "flex",
  "height": "100px",
  "width": "100px",
}
`;

exports[`Styling and specificity Dynamic imports - style-loader on mobile theme A should be styled correctly 1`] = `
Object {
  "background-color": "#ff0",
  "display": "flex",
  "height": "100px",
  "width": "100px",
}
`;

exports[`Styling and specificity Dynamic imports - style-loader on mobile theme B should be styled correctly 1`] = `
Object {
  "background-color": "red",
  "display": "flex",
  "height": "100px",
  "width": "100px",
}
`;

exports[`Styling and specificity Simple - development mode (style-loader) -  on desktop and hovered should be styled correctly 1`] = `
Object {
  "background-color": "purple",
  "color": "green",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Simple - development mode (style-loader) -  on desktop should be styled correctly 1`] = `
Object {
  "color": "green",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Simple - development mode (style-loader) -  on mobile should be styled correctly 1`] = `
Object {
  "color": "red",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Simple - mini-css-extract-loader -  on desktop and hovered should be styled correctly 1`] = `
Object {
  "background-color": "purple",
  "color": "green",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Simple - mini-css-extract-loader -  on desktop should be styled correctly 1`] = `
Object {
  "color": "green",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Simple - mini-css-extract-loader -  on mobile should be styled correctly 1`] = `
Object {
  "color": "red",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Simple - style-loader -  on desktop and hovered should be styled correctly 1`] = `
Object {
  "background-color": "purple",
  "color": "green",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Simple - style-loader -  on desktop should be styled correctly 1`] = `
Object {
  "color": "green",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Simple - style-loader -  on mobile should be styled correctly 1`] = `
Object {
  "color": "red",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Themed - development mode (style-loader) should be styled correctly with theme one 1`] = `
Object {
  "background-color": "yellow",
  "border": "6px solid pink",
  "border-bottom-color": "pink",
  "border-bottom-style": "solid",
  "border-bottom-width": "6px",
  "border-image-outset": "initial",
  "border-image-repeat": "initial",
  "border-image-slice": "initial",
  "border-image-source": "initial",
  "border-image-width": "initial",
  "border-left-color": "pink",
  "border-left-style": "solid",
  "border-left-width": "6px",
  "border-right-color": "pink",
  "border-right-style": "solid",
  "border-right-width": "6px",
  "border-top-color": "pink",
  "border-top-style": "solid",
  "border-top-width": "6px",
  "display": "flex",
  "height": "12px",
  "line-height": "6",
  "z-index": "3",
}
`;

exports[`Styling and specificity Themed - development mode (style-loader) should be styled correctly with theme two 1`] = `
Object {
  "background-color": "red",
  "border": "8px solid pink",
  "border-bottom-color": "pink",
  "border-bottom-style": "solid",
  "border-bottom-width": "8px",
  "border-image-outset": "initial",
  "border-image-repeat": "initial",
  "border-image-slice": "initial",
  "border-image-source": "initial",
  "border-image-width": "initial",
  "border-left-color": "pink",
  "border-left-style": "solid",
  "border-left-width": "8px",
  "border-right-color": "pink",
  "border-right-style": "solid",
  "border-right-width": "8px",
  "border-top-color": "pink",
  "border-top-style": "solid",
  "border-top-width": "8px",
  "display": "flex",
  "height": "8px",
  "line-height": "4",
  "z-index": "3",
}
`;

exports[`Styling and specificity Themed - mini-css-extract-loader should be styled correctly with theme one 1`] = `
Object {
  "background-color": "#ff0",
  "border": "6px solid pink",
  "border-bottom-color": "pink",
  "border-bottom-style": "solid",
  "border-bottom-width": "6px",
  "border-image-outset": "initial",
  "border-image-repeat": "initial",
  "border-image-slice": "initial",
  "border-image-source": "initial",
  "border-image-width": "initial",
  "border-left-color": "pink",
  "border-left-style": "solid",
  "border-left-width": "6px",
  "border-right-color": "pink",
  "border-right-style": "solid",
  "border-right-width": "6px",
  "border-top-color": "pink",
  "border-top-style": "solid",
  "border-top-width": "6px",
  "display": "flex",
  "height": "12px",
  "line-height": "6",
  "z-index": "3",
}
`;

exports[`Styling and specificity Themed - mini-css-extract-loader should be styled correctly with theme two 1`] = `
Object {
  "background-color": "red",
  "border": "8px solid pink",
  "border-bottom-color": "pink",
  "border-bottom-style": "solid",
  "border-bottom-width": "8px",
  "border-image-outset": "initial",
  "border-image-repeat": "initial",
  "border-image-slice": "initial",
  "border-image-source": "initial",
  "border-image-width": "initial",
  "border-left-color": "pink",
  "border-left-style": "solid",
  "border-left-width": "8px",
  "border-right-color": "pink",
  "border-right-style": "solid",
  "border-right-width": "8px",
  "border-top-color": "pink",
  "border-top-style": "solid",
  "border-top-width": "8px",
  "display": "flex",
  "height": "8px",
  "line-height": "4",
  "z-index": "3",
}
`;

exports[`Styling and specificity Themed - style-loader should be styled correctly with theme one 1`] = `
Object {
  "background-color": "#ff0",
  "border": "6px solid pink",
  "border-bottom-color": "pink",
  "border-bottom-style": "solid",
  "border-bottom-width": "6px",
  "border-image-outset": "initial",
  "border-image-repeat": "initial",
  "border-image-slice": "initial",
  "border-image-source": "initial",
  "border-image-width": "initial",
  "border-left-color": "pink",
  "border-left-style": "solid",
  "border-left-width": "6px",
  "border-right-color": "pink",
  "border-right-style": "solid",
  "border-right-width": "6px",
  "border-top-color": "pink",
  "border-top-style": "solid",
  "border-top-width": "6px",
  "display": "flex",
  "height": "12px",
  "line-height": "6",
  "z-index": "3",
}
`;

exports[`Styling and specificity Themed - style-loader should be styled correctly with theme two 1`] = `
Object {
  "background-color": "red",
  "border": "8px solid pink",
  "border-bottom-color": "pink",
  "border-bottom-style": "solid",
  "border-bottom-width": "8px",
  "border-image-outset": "initial",
  "border-image-repeat": "initial",
  "border-image-slice": "initial",
  "border-image-source": "initial",
  "border-image-width": "initial",
  "border-left-color": "pink",
  "border-left-style": "solid",
  "border-left-width": "8px",
  "border-right-color": "pink",
  "border-right-style": "solid",
  "border-right-width": "8px",
  "border-top-color": "pink",
  "border-top-style": "solid",
  "border-top-width": "8px",
  "display": "flex",
  "height": "8px",
  "line-height": "4",
  "z-index": "3",
}
`;

exports[`Styling and specificity Unused modules - development mode (style-loader) should be styled correctly 1`] = `
Object {
  "display": "block",
  "height": "100px",
  "width": "100px",
}
`;

exports[`Styling and specificity Unused modules - mini-css-extract-loader should be styled correctly 1`] = `
Object {
  "display": "block",
  "height": "100px",
  "width": "100px",
}
`;

exports[`Styling and specificity Unused modules - style-loader should be styled correctly 1`] = `
Object {
  "display": "block",
  "height": "100px",
  "width": "100px",
}
`;
